mono/profiler/arg-parsing.c

```diff
@@ -0,0 +1,406 @@
+		const char *end = strchr (arg, '=');
+
+		*rval = NULL;
+		if (!end)
+			return !strcmp (arg, opt_name);
+
+		if (strncmp (arg, opt_name, strlen (opt_name)) || (end - arg) > strlen (opt_name) + 1)
+			return FALSE;
+		*rval = end + 1;
+		return TRUE;
+	} else {
+		//FIXME how should we handle passing a value to an arg that doesn't expect it?
+		return !strcmp (arg, opt_name);
+	}
+}
+
+static void
+parse_arg (const char *arg, ProfilerConfig *config)
+{
+	const char *val;
+
+	if (match_option (arg, "help", NULL)) {
+		usage ();
+	} else if (match_option (arg, "report", NULL)) {
+		config->do_report = TRUE;
+	} else if (match_option (arg, "debug", NULL)) {
+		config->do_debug = TRUE;
+	} else if (match_option (arg, "debug-coverage", NULL)) {
+		config->debug_coverage = TRUE;
+	} else if (match_option (arg, "sampling-real", NULL)) {
+		config->sampling_mode = MONO_PROFILER_STAT_MODE_REAL;
+	} else if (match_option (arg, "sampling-process", NULL)) {
+		config->sampling_mode = MONO_PROFILER_STAT_MODE_PROCESS;
+	} else if (match_option (arg, "heapshot", &val)) {
+		config->enable_mask |= HeapShotAlias;
+		set_hsmode (config, val);
+	} else if (match_option (arg, "sample", &val)) {
+		config->enable_mask |= PerfSamplingAlias;
+		set_sample_freq (config, val);
+	} else if (match_option (arg, "zip", NULL)) {
+		config->use_zip = TRUE;
+	} else if (match_option (arg, "output", &val)) {
+		config->output_filename = g_strdup (val);
+	} else if (match_option (arg, "port", &val)) {
+		char *end;
+		config->command_port = strtoul (val, &end, 10);
+	} else if (match_option (arg, "maxframes", &val)) {
+		char *end;
+		int num_frames = strtoul (val, &end, 10);
+		if (num_frames > MAX_FRAMES)
+			num_frames = MAX_FRAMES;
+		config->notraces = num_frames == 0;
+		config->num_frames = num_frames;
+	} else if (match_option (arg, "maxsamples", &val)) {
+		char *end;
+		int max_samples = strtoul (val, &end, 10);
+		if (max_samples)
+			config->max_allocated_sample_hits = max_samples;
+	} else if (match_option (arg, "calldepth", &val)) {
+		char *end;
+		config->max_call_depth = strtoul (val, &end, 10);
+	} else if (match_option (arg, "covfilter-file", &val)) {
+		if (config->cov_filter_files == NULL)
+			config->cov_filter_files = g_ptr_array_new ();
+		g_ptr_array_add (config->cov_filter_files, g_strdup (val));
+	} else if (match_option (arg, "onlycoverage", NULL)) {
+		config->only_coverage = TRUE;
+	} else {
+		int i;
+
+		for (i = 0; i < G_N_ELEMENTS (event_list); ++i){
+			if (!strcmp (arg, event_list [i].event_name)) {
+				config->enable_mask |= event_list [i].mask;
+				break;
+			} else if (arg [0] == 'n' && arg [1] == 'o' && !strcmp (arg + 2, event_list [i].event_name)) {
+				config->disable_mask |= event_list [i].mask;
+			}
+		}
+		if (i == G_N_ELEMENTS (event_list)) {
+			printf ("Could not parse argument %s\n", arg);
+		}
+	}
+}
+
+static void
+load_args_from_env_or_default (ProfilerConfig *config)
+{
+	//XXX change this to header constants
```

Comment:
Consistency with the rest of the runtime.

All constants are better thrown together in a conf header. Similar to sgen-conf.h

It allow us to have long form description/discussion of their values.